File: src/main/java/org/cubexmc/metro/util/LocationUtil.java
package org.cubexmc.metro.util;

import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.block.data.BlockData;
import org.bukkit.block.data.Powerable;
import org.bukkit.block.data.Rail;
import org.bukkit.util.Vector;

import java.util.EnumSet;
import java.util.Set;

public class LocationUtil {

    public enum RailType {
        STRAIGHT,
        CURVE,
        ASCENDING,
        DESCENDING,
        UNKNOWN
    }

    private static final Set<Material> RAIL_MATERIALS = EnumSet.of(
        Material.RAIL,
        Material.POWERED_RAIL,
        Material.DETECTOR_RAIL,
        Material.ACTIVATOR_RAIL
    );

    public static String locationToString(Location location) {
        if (location == null || location.getWorld() == null) return null;
        return String.format("%s,%d,%d,%d",
                location.getWorld().getName(),
                location.getBlockX(),
                location.getBlockY(),
                location.getBlockZ());
    }

    public static Location locationFromString(String world, String locationString) {
        if (locationString == null) return null;
        String[] parts = locationString.split(",");
        if (parts.length != 4) return null;
        try {
            double x = Double.parseDouble(parts[1]);
            double y = Double.parseDouble(parts[2]);
            double z = Double.parseDouble(parts[3]);
            return new Location(
                    org.bukkit.Bukkit.getWorld(world != null ? world : parts[0]),
                    x, y, z);
        } catch (NumberFormatException e) {
            return null;
        }
    }

    public static boolean isSameLocation(Location loc1, Location loc2) {
        if (loc1 == null || loc2 == null) return false;
        return loc1.getWorld() != null && loc1.getWorld().equals(loc2.getWorld())
                && loc1.getBlockX() == loc2.getBlockX()
                && loc1.getBlockY() == loc2.getBlockY()
                && loc1.getBlockZ() == loc2.getBlockZ();
    }

    public static Vector getDirectionVector(Location from, Location to) {
        return to.toVector().subtract(from.toVector()).normalize();
    }

    public static boolean isRail(Location location) {
        if (location == null) return false;
        Block block = location.getBlock();
        if (isRailBlock(block.getType())) return true;
        return isRailBlock(block.getRelative(0, -1, 0).getType());
    }

    public static boolean isOnRail(Location location) {
        return isRail(location);
    }

    private static boolean isRailBlock(Material material) {
        return material != null && material.name().contains("RAIL");
    }

    public static boolean isPoweredAscendingRailPowered(Location location) {
        if (location == null) return false;
        Rail rail = getRailData(location.getBlock());
        if (rail == null) rail = getRailData(location.getBlock().getRelative(0, -1, 0));
        if (rail == null) return false;
        Rail.Shape shape = rail.getShape();
        boolean ascending = shape == Rail.Shape.ASCENDING_EAST || shape == Rail.Shape.ASCENDING_WEST
                || shape == Rail.Shape.ASCENDING_NORTH || shape == Rail.Shape.ASCENDING_SOUTH;
        if (!ascending) return false;
        if (rail instanceof Powerable) return ((Powerable) rail).isPowered();
        return false;
    }

    private static Rail getRailData(Block block) {
        if (block == null) return null;
        BlockData data = block.getBlockData();
        return data instanceof Rail ? (Rail) data : null;
    }

    public static boolean isCurveRail(Location location) {
        Rail rail = getRailData(location != null ? location.getBlock() : null);
        if (rail == null) rail = getRailData(location != null ? location.getBlock().getRelative(0, -1, 0) : null);
        if (rail == null) return false;
        switch (rail.getShape()) {
            case NORTH_EAST: case NORTH_WEST: case SOUTH_EAST: case SOUTH_WEST:
                return true;
            default:
                return false;
        }
    }

    private static Location computeRailCenter(Block block, Rail rail) {
        Location center = new Location(block.getWorld(), block.getX() + 0.5, block.getY(), block.getZ() + 0.5);
        switch (rail.getShape()) {
            case ASCENDING_EAST: case ASCENDING_WEST: case ASCENDING_NORTH: case ASCENDING_SOUTH:
                center.add(0.0, 0.5, 0.0);
                break;
            default:
                break;
        }
        return center;
    }

    private static Location centerRail(Location candidate) {
        if (candidate == null) return candidate;
        Block block = candidate.getBlock();
        Rail rail = getRailData(block);
        if (rail != null) return computeRailCenter(block, rail);
        Block below = block.getRelative(0, -1, 0);
        rail = getRailData(below);
        if (rail != null) return computeRailCenter(below, rail);
        return center(candidate);
    }

    public static RailType getRailType(Location location) {
        if (location == null) return RailType.UNKNOWN;
        Block block = location.getBlock();
        BlockData data = block.getBlockData();
        if (!(data instanceof Rail)) {
            Block below = block.getRelative(0, -1, 0);
            if (below.getBlockData() instanceof Rail) data = below.getBlockData();
            else return RailType.UNKNOWN;
        }
        Rail rail = (Rail) data;
        switch (rail.getShape()) {
            case NORTH_SOUTH: case EAST_WEST: return RailType.STRAIGHT;
            case NORTH_EAST: case NORTH_WEST: case SOUTH_EAST: case SOUTH_WEST: return RailType.CURVE;
            case ASCENDING_EAST: case ASCENDING_WEST: case ASCENDING_NORTH: case ASCENDING_SOUTH: return RailType.ASCENDING;
            default: return RailType.UNKNOWN;
        }
    }

    public static double getSafeSpeedForRail(RailType type, double baseSpeed, boolean safeMode) {
        if (!safeMode) return baseSpeed;
        switch (type) {
            case STRAIGHT: return baseSpeed;
            case CURVE: return Math.min(baseSpeed * 0.65, 0.25);
            case ASCENDING: return Math.min(baseSpeed * 0.87, 0.35);
            case DESCENDING: return Math.min(baseSpeed * 0.95, 0.42);
            default: return Math.min(baseSpeed * 0.80, 0.32);
        }
    }

    public static Vector vectorFromYaw(float yaw) {
        double radians = Math.toRadians(yaw);
        double x = -Math.sin(radians);
        double z = Math.cos(radians);
        Vector vector = new Vector(x, 0, z);
        return vector.lengthSquared() == 0 ? new Vector(0, 0, 0) : vector.normalize();
    }

    public static Location center(Location location) {
        if (location == null) return null;
        double x = location.getX(), z = location.getZ();
        if (Math.abs(x - Math.floor(x)) < 1e-6 && Math.abs(z - Math.floor(z)) < 1e-6) {
            return location.clone().add(0.5, 0, 0.5);
        }
        return location.clone();
    }

    public static Location snapToRail(Location location, org.bukkit.World world) {
        if (location == null || world == null) return location;
        Location best = null;
        double bestDist = Double.MAX_VALUE;
        int bx = location.getBlockX(), by = location.getBlockY(), bz = location.getBlockZ();
        for (int y = -1; y <= 1; y++)
            for (int x = -1; x <= 1; x++)
                for (int z = -1; z <= 1; z++) {
                    Location c = new Location(world, bx + x, by + y, bz + z);
                    if (isRail(c)) {
                        Location center = centerRail(c);
                        double d = center.distanceSquared(location);
                        if (d < bestDist) { bestDist = d; best = center; }
                    }
                }
        return best != null ? best : location;
    }

    public static Location findNearestRail(Location location, double radius) {
        if (location == null || location.getWorld() == null) return null;
        Location best = null;
        double bestDistSq = radius * radius;
        int r = (int) Math.ceil(radius), bx = location.getBlockX(), by = location.getBlockY(), bz = location.getBlockZ();
        for (int dx = -r; dx <= r; dx++)
            for (int dy = -r; dy <= r; dy++)
                for (int dz = -r; dz <= r; dz++) {
                    Location c = new Location(location.getWorld(), bx + dx, by + dy, bz + dz);
                    if (isRail(c)) {
                        Location center = centerRail(c);
                        double ds = center.distanceSquared(location);
                        if (ds < bestDistSq) { bestDistSq = ds; best = center; }
                    }
                }
        return best;
    }

    public static Vector railDirection(Location location, Vector fallback) {
        if (location == null) return fallback;
        Block block = location.getBlock();
        BlockData data = block.getBlockData();
        Rail rail = null;
        if (data instanceof Rail) rail = (Rail) data;
        else {
            Block below = block.getRelative(0, -1, 0);
            if (below.getBlockData() instanceof Rail) rail = (Rail) below.getBlockData();
        }
        if (rail != null) {
            Vector dir = vectorFromRailShape(rail.getShape(), location, fallback);
            if (dir != null) {
                Rail.Shape shape = rail.getShape();
                if (shape != Rail.Shape.NORTH_EAST && shape != Rail.Shape.NORTH_WEST
                        && shape != Rail.Shape.SOUTH_EAST && shape != Rail.Shape.SOUTH_WEST) {
                    if (fallback != null && dir.dot(fallback) < 0) dir.multiply(-1);
                }
                return dir;
            }
        }
        return fallback;
    }

    private static Vector vectorFromRailShape(Rail.Shape shape, Location location, Vector fallback) {
        if (shape == null) return null;
        switch (shape) {
            case NORTH_SOUTH: return new Vector(0, 0, 1);
            case EAST_WEST: return new Vector(1, 0, 0);
            case ASCENDING_EAST: return new Vector(1, 1, 0);
            case ASCENDING_WEST: return new Vector(-1, 1, 0);
            case ASCENDING_NORTH: return new Vector(0, 1, -1);
            case ASCENDING_SOUTH: return new Vector(0, 1, 1);
            case SOUTH_EAST:
                return computeCurveDirection(location, fallback, new Vector(0, 0, 1), new Vector(1, 0, 0));
            case SOUTH_WEST:
                return computeCurveDirection(location, fallback, new Vector(0, 0, 1), new Vector(-1, 0, 0));
            case NORTH_WEST:
                return computeCurveDirection(location, fallback, new Vector(0, 0, -1), new Vector(-1, 0, 0));
            case NORTH_EAST:
                return computeCurveDirection(location, fallback, new Vector(0, 0, -1), new Vector(1, 0, 0));
            default: return null;
        }
    }

    private static Vector computeCurveDirection(Location location, Vector fallback, Vector dir1, Vector dir2) {
        double dx = location.getX() - Math.floor(location.getX()) - 0.5;
        double dz = location.getZ() - Math.floor(location.getZ()) - 0.5;
        Vector radial;
        if (Math.abs(dx) < 1e-3 && Math.abs(dz) < 1e-3) {
            Vector fb = (fallback != null) ? fallback.clone() : new Vector(1, 0, 0);
            if (fb.lengthSquared() < 1e-6) fb = new Vector(1, 0, 0);
            fb.setY(0); fb.normalize();
            radial = new Vector(-fb.getX(), 0, -fb.getZ());
        } else {
            radial = new Vector(dx, 0, dz);
        }
        Vector cw = new Vector(radial.getZ(), 0, -radial.getX());
        Vector ccw = new Vector(-radial.getZ(), 0, radial.getX());
        if (fallback == null || fallback.lengthSquared() < 1e-6) {
            Vector avg = dir1.clone().add(dir2).normalize();
            return (cw.dot(avg) >= ccw.dot(avg) ? cw : ccw).normalize();
        }
        return (cw.dot(fallback) >= ccw.dot(fallback) ? cw : ccw).normalize();
    }
}

File: src/main/java/org/cubexmc/metro/util/LocationUtil.kt
package org.cubexmc.metro.util

import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.World
import org.bukkit.block.Block
import org.bukkit.block.data.Powerable
import org.bukkit.block.data.Rail
import org.bukkit.util.Vector
import java.util.EnumSet
import kotlin.math.abs
import kotlin.math.ceil
import kotlin.math.cos
import kotlin.math.floor
import kotlin.math.min
import kotlin.math.sin

object LocationUtil {

    enum class RailType {
        STRAIGHT,
        CURVE,
        ASCENDING,
        DESCENDING,
        UNKNOWN,
    }

    private const val EPSILON = 1e-6
    private const val CURVE_CENTER_EPSILON = 1e-3
    private const val BLOCK_CENTER_OFFSET = 0.5

    @Suppress("unused")
    private val RAIL_MATERIALS: Set<Material> = EnumSet.of(
        Material.RAIL,
        Material.POWERED_RAIL,
        Material.DETECTOR_RAIL,
        Material.ACTIVATOR_RAIL,
    )

    @JvmStatic
    fun locationToString(location: Location?): String? {
        val world = location?.world ?: return null
        return String.format(
            "%s,%d,%d,%d",
            world.name,
            location.blockX,
            location.blockY,
            location.blockZ,
        )
    }

    @JvmStatic
    fun locationFromString(world: String?, locationString: String?): Location? {
        if (locationString == null) return null
        val parts = locationString.split(",")
        if (parts.size != 4) return null
        return try {
            val x = parts[1].toDouble()
            val y = parts[2].toDouble()
            val z = parts[3].toDouble()
            Location(Bukkit.getWorld(world ?: parts[0]), x, y, z)
        } catch (_: NumberFormatException) {
            null
        }
    }

    @JvmStatic
    fun isSameLocation(loc1: Location?, loc2: Location?): Boolean {
        if (loc1 == null || loc2 == null) return false
        val world = loc1.world ?: return false
        return world == loc2.world &&
            loc1.blockX == loc2.blockX &&
            loc1.blockY == loc2.blockY &&
            loc1.blockZ == loc2.blockZ
    }

    @JvmStatic
    fun getDirectionVector(from: Location, to: Location): Vector =
        to.toVector().subtract(from.toVector()).normalize()

    @JvmStatic
    fun isRail(location: Location?): Boolean {
        if (location == null) return false
        val block = location.block
        if (isRailBlock(block.type)) return true
        // Java 版本对 getRelative 的结果做过 null 判断（测试里的 mock Block 会返回 null），保持同样的宽容
        val below = relativeBlockOrNull(block, 0, -1, 0) ?: return false
        return isRailBlock(below.type)
    }

    /** Bukkit 标注 getRelative 为非空，这里刻意放宽成可空以保留 Java 版本的防御性判断。 */
    private fun relativeBlockOrNull(block: Block, x: Int, y: Int, z: Int): Block? = block.getRelative(x, y, z)

    @JvmStatic
    fun isOnRail(location: Location?): Boolean = isRail(location)

    private fun isRailBlock(material: Material?): Boolean = material != null && material.name.contains("RAIL")

    @JvmStatic
    fun isPoweredAscendingRailPowered(location: Location?): Boolean {
        if (location == null) return false
        val rail = getRailData(location.block) ?: getRailData(location.block.getRelative(0, -1, 0)) ?: return false
        val shape = rail.shape
        val ascending = shape == Rail.Shape.ASCENDING_EAST ||
            shape == Rail.Shape.ASCENDING_WEST ||
            shape == Rail.Shape.ASCENDING_NORTH ||
            shape == Rail.Shape.ASCENDING_SOUTH
        if (!ascending) return false
        return rail is Powerable && (rail as Powerable).isPowered
    }

    private fun getRailData(block: Block?): Rail? {
        if (block == null) return null
        val data = block.blockData
        return data as? Rail
    }

    @JvmStatic
    fun isCurveRail(location: Location?): Boolean {
        val rail = getRailData(location?.block)
            ?: getRailData(location?.block?.getRelative(0, -1, 0))
            ?: return false
        return when (rail.shape) {
            Rail.Shape.NORTH_EAST, Rail.Shape.NORTH_WEST, Rail.Shape.SOUTH_EAST, Rail.Shape.SOUTH_WEST -> true
            else -> false
        }
    }

    private fun computeRailCenter(block: Block, rail: Rail): Location {
        val center = Location(
            block.world,
            block.x + BLOCK_CENTER_OFFSET,
            block.y.toDouble(),
            block.z + BLOCK_CENTER_OFFSET,
        )
        when (rail.shape) {
            Rail.Shape.ASCENDING_EAST,
            Rail.Shape.ASCENDING_WEST,
            Rail.Shape.ASCENDING_NORTH,
            Rail.Shape.ASCENDING_SOUTH,
            -> center.add(0.0, BLOCK_CENTER_OFFSET, 0.0)

            else -> Unit
        }
        return center
    }

    private fun centerRail(candidate: Location?): Location? {
        if (candidate == null) return null
        val block = candidate.block
        val rail = getRailData(block)
        if (rail != null) return computeRailCenter(block, rail)
        val below = relativeBlockOrNull(block, 0, -1, 0)
        val belowRail = getRailData(below)
        if (below != null && belowRail != null) return computeRailCenter(below, belowRail)
        return center(candidate)
    }

    @JvmStatic
    fun getRailType(location: Location?): RailType {
        if (location == null) return RailType.UNKNOWN
        val block = location.block
        var data = block.blockData
        if (data !is Rail) {
            val below = block.getRelative(0, -1, 0)
            val belowData = below.blockData
            if (belowData is Rail) {
                data = belowData
            } else {
                return RailType.UNKNOWN
            }
        }
        return when ((data as Rail).shape) {
            Rail.Shape.NORTH_SOUTH, Rail.Shape.EAST_WEST -> RailType.STRAIGHT
            Rail.Shape.NORTH_EAST,
            Rail.Shape.NORTH_WEST,
            Rail.Shape.SOUTH_EAST,
            Rail.Shape.SOUTH_WEST,
            -> RailType.CURVE

            Rail.Shape.ASCENDING_EAST,
            Rail.Shape.ASCENDING_WEST,
            Rail.Shape.ASCENDING_NORTH,
            Rail.Shape.ASCENDING_SOUTH,
            -> RailType.ASCENDING

            else -> RailType.UNKNOWN
        }
    }

    @JvmStatic
    @Suppress("MagicNumber")
    fun getSafeSpeedForRail(type: RailType, baseSpeed: Double, safeMode: Boolean): Double {
        if (!safeMode) return baseSpeed
        return when (type) {
            RailType.STRAIGHT -> baseSpeed
            RailType.CURVE -> min(baseSpeed * 0.65, 0.25)
            RailType.ASCENDING -> min(baseSpeed * 0.87, 0.35)
            RailType.DESCENDING -> min(baseSpeed * 0.95, 0.42)
            else -> min(baseSpeed * 0.80, 0.32)
        }
    }

    @JvmStatic
    fun vectorFromYaw(yaw: Float): Vector {
        val radians = Math.toRadians(yaw.toDouble())
        val x = -sin(radians)
        val z = cos(radians)
        val vector = Vector(x, 0.0, z)
        return if (vector.lengthSquared() == 0.0) Vector(0, 0, 0) else vector.normalize()
    }

    @JvmStatic
    fun center(location: Location?): Location? {
        if (location == null) return null
        val x = location.x
        val z = location.z
        if (abs(x - floor(x)) < EPSILON && abs(z - floor(z)) < EPSILON) {
            return location.clone().add(BLOCK_CENTER_OFFSET, 0.0, BLOCK_CENTER_OFFSET)
        }
        return location.clone()
    }

    @JvmStatic
    fun snapToRail(location: Location?, world: World?): Location? {
        if (location == null || world == null) return location
        var best: Location? = null
        var bestDist = Double.MAX_VALUE
        val bx = location.blockX
        val by = location.blockY
        val bz = location.blockZ
        for (y in -1..1) {
            for (x in -1..1) {
                for (z in -1..1) {
                    val candidate = Location(world, (bx + x).toDouble(), (by + y).toDouble(), (bz + z).toDouble())
                    if (!isRail(candidate)) continue
                    val center = centerRail(candidate) ?: continue
                    val distance = center.distanceSquared(location)
                    if (distance < bestDist) {
                        bestDist = distance
                        best = center
                    }
                }
            }
        }
        return best ?: location
    }

    @JvmStatic
    fun findNearestRail(location: Location?, radius: Double): Location? {
        val world = location?.world ?: return null
        var best: Location? = null
        var bestDistSq = radius * radius
        val r = ceil(radius).toInt()
        val bx = location.blockX
        val by = location.blockY
        val bz = location.blockZ
        for (dx in -r..r) {
            for (dy in -r..r) {
                for (dz in -r..r) {
                    val candidate =
                        Location(world, (bx + dx).toDouble(), (by + dy).toDouble(), (bz + dz).toDouble())
                    if (!isRail(candidate)) continue
                    val center = centerRail(candidate) ?: continue
                    val distance = center.distanceSquared(location)
                    if (distance < bestDistSq) {
                        bestDistSq = distance
                        best = center
                    }
                }
            }
        }
        return best
    }

    @JvmStatic
    fun railDirection(location: Location?, fallback: Vector?): Vector? {
        if (location == null) return fallback
        val block = location.block
        val data = block.blockData
        val rail = data as? Rail ?: block.getRelative(0, -1, 0).blockData as? Rail
        if (rail != null) {
            val dir = vectorFromRailShape(rail.shape, location, fallback)
            if (dir != null) {
                val shape = rail.shape
                if (shape != Rail.Shape.NORTH_EAST && shape != Rail.Shape.NORTH_WEST &&
                    shape != Rail.Shape.SOUTH_EAST && shape != Rail.Shape.SOUTH_WEST
                ) {
                    if (fallback != null && dir.dot(fallback) < 0) dir.multiply(-1)
                }
                return dir
            }
        }
        return fallback
    }

    private fun vectorFromRailShape(shape: Rail.Shape?, location: Location, fallback: Vector?): Vector? {
        if (shape == null) return null
        return when (shape) {
            Rail.Shape.NORTH_SOUTH -> Vector(0, 0, 1)
            Rail.Shape.EAST_WEST -> Vector(1, 0, 0)
            Rail.Shape.ASCENDING_EAST -> Vector(1, 1, 0)
            Rail.Shape.ASCENDING_WEST -> Vector(-1, 1, 0)
            Rail.Shape.ASCENDING_NORTH -> Vector(0, 1, -1)
            Rail.Shape.ASCENDING_SOUTH -> Vector(0, 1, 1)
            Rail.Shape.SOUTH_EAST ->
                computeCurveDirection(location, fallback, Vector(0, 0, 1), Vector(1, 0, 0))

            Rail.Shape.SOUTH_WEST ->
                computeCurveDirection(location, fallback, Vector(0, 0, 1), Vector(-1, 0, 0))

            Rail.Shape.NORTH_WEST ->
                computeCurveDirection(location, fallback, Vector(0, 0, -1), Vector(-1, 0, 0))

            Rail.Shape.NORTH_EAST ->
                computeCurveDirection(location, fallback, Vector(0, 0, -1), Vector(1, 0, 0))
        }
    }

    private fun computeCurveDirection(
        location: Location,
        fallback: Vector?,
        dir1: Vector,
        dir2: Vector,
    ): Vector {
        val dx = location.x - floor(location.x) - BLOCK_CENTER_OFFSET
        val dz = location.z - floor(location.z) - BLOCK_CENTER_OFFSET
        val radial: Vector
        if (abs(dx) < CURVE_CENTER_EPSILON && abs(dz) < CURVE_CENTER_EPSILON) {
            var fb = fallback?.clone() ?: Vector(1, 0, 0)
            if (fb.lengthSquared() < EPSILON) fb = Vector(1, 0, 0)
            fb.y = 0.0
            fb.normalize()
            radial = Vector(-fb.x, 0.0, -fb.z)
        } else {
            radial = Vector(dx, 0.0, dz)
        }
        val cw = Vector(radial.z, 0.0, -radial.x)
        val ccw = Vector(-radial.z, 0.0, radial.x)
        if (fallback == null || fallback.lengthSquared() < EPSILON) {
            val avg = dir1.clone().add(dir2).normalize()
            return (if (cw.dot(avg) >= ccw.dot(avg)) cw else ccw).normalize()
        }
        return (if (cw.dot(fallback) >= ccw.dot(fallback)) cw else ccw).normalize()
    }
}
